src/main/scala/io/iohk/ethereum/ledger/Ledger.scala

```diff
@@ -1,59 +1,319 @@
+
+    val staleBlocksWithReceiptsAndTDs = removeBlocksUntil(parent, bestNumber).reverse
+    val staleBlocks = staleBlocksWithReceiptsAndTDs.map(_._1)
+
+    for (block <- staleBlocks) yield blockQueue.enqueueBlock(block)
+
+    val (executedBlocks, maybeError) = executeBlocks(newBranch, parentTd)
+    maybeError match {
+      case None =>
+        Right(staleBlocks, executedBlocks)
+
+      case Some(error) =>
+        revertChainReorganisation(newBranch, staleBlocksWithReceiptsAndTDs, executedBlocks)
+        Left(error)
+    }
+  }
+
+  /**
+    * Used to revert chain reorganisation in the event that one of the blocks from new branch
+    * fails to execute
+    *
+    * @param newBranch - new blocks
+    * @param oldBranch - old blocks along with corresponding receipts and totalDifficulties
+    * @param executedBlocks - sub-sequence of new branch that was executed correctly
+    */
+  private def revertChainReorganisation(newBranch: List[Block], oldBranch: List[(Block, Seq[Receipt], BigInt)],
+    executedBlocks: List[Block]): Unit = {
+
+    if (executedBlocks.nonEmpty) {
+      removeBlocksUntil(executedBlocks.head.header.parentHash, executedBlocks.last.header.number)
+    }
+
+    oldBranch.foreach { case (block, receipts, td) =>
+      blockchain.save(block, receipts, td, saveAsBestBlock = false)
+    }
+
+    val bestNumber = oldBranch.last._1.header.number
+    blockchain.saveBestBlockNumber(bestNumber)
+    executedBlocks.foreach(blockQueue.enqueueBlock(_, bestNumber))
+
+    newBranch.diff(executedBlocks).headOption.foreach { block =>
+      blockQueue.removeSubtree(block.header.hash)
+    }
+  }
+
+  /**
+    * Executes a list blocks, storing the results in the blockchain
+    * @param blocks block to be executed
+    * @return a list of blocks that were correctly executed and an optional [[BlockExecutionError]]
+    */
+  private def executeBlocks(blocks: List[Block], parentTd: BigInt): (List[Block], Option[BlockExecutionError]) = {
+    blocks match {
+      case block :: remainingBlocks =>
+        executeBlock (block) match {
+          case Right (receipts) =>
+            val td = parentTd + block.header.difficulty
+            blockchain.save(block, receipts, td, saveAsBestBlock = true)
+
+            val (executedBlocks, error) = executeBlocks(remainingBlocks, td)
+            (block :: executedBlocks, error)
+
+          case Left(error) =>
+          (Nil, Some(error))
+        }
+
+      case Nil =>
+        (Nil, None)
+    }
+  }
+
+  /**
+    * Remove blocks from the [[Blockchain]] along with receipts and total difficulties
+    * @param parent remove blocks until this hash (exclusive)
+    * @param fromNumber start removing from this number (downwards)
+    * @return the list of removed blocks along with receipts and total difficulties
+    */
+  private def removeBlocksUntil(parent: ByteString, fromNumber: BigInt): List[(Block, Seq[Receipt], BigInt)] = {
```

Comment:
I think we might have an issue here. In this function we are removing blocks from blockchain but we are not "undoing" state changes.

Also, while reviewing the code, @ntallar discovered (and I agree after discussion) that current pruning implementation has a bug that will cause to remove valid state nodes after reorg (specific details are no worth to be mentioned here imo). I think we will need to define a way to let the pruning know that we are doing a blockchain reorg.

Reply:
Yes, that was intentional. Even if we didn't prune those extra nodes, thy shouldn't be a problem.
But if valid nodes may be removed that's a different story... I'm keen on getting to know the details, and what we can do about it...

Reply:
Currently we have a super naive pruning approach based on MptNode reference count: 
- We increment when someone inserts it and decrement when it's removed
- We track the block number when mpt node reference count reaches to zero, lets call it `Bn`
- We define a parameter `Offset` we will "hold" that node
- if at `Bn` + `Offset` references are still 0, we remove mpt node as it might not be used by anyone else

**The problem is that we are not updating reference count when dropping block executions after a reorg (actually we would need to change internal implementation but that shouldn't be a problem from the Ledger point of view)**

The following example demonstrates why we need to let know pruning algo that we are undoing a block:
- `Offset` = 5
- Let's suppose that we execute the blocks in the following order: b1, b2, b3, b4, b5', [Reorg], b5, b6, b7, b8, b9, b10
- At b2, NodeX references is 1
- At b5' NodeX references is 0 and marked as prune candidate
- At b10 we will remove NodeX as it still has 0 references but we shouldn't because b5' changes do not count after doing a reorg so reference count should be 1

![prune_example](https://user-images.githubusercontent.com/3750504/31544886-566a8c4e-aff3-11e7-9f79-d71c96b9f552.png)

Reply:
Thanks for this explanation, indeed we have a problem. However it seems it's not a new problem, related to this PR, as we already are doing re-orgs [by removing blocks](https://github.com/input-output-hk/mantis/blob/phase/daedalus/src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSync.scala#L182) 🙁

Reply:
@rtkaczyk yes, the bug was present before this PR. The good thing is:
- It forced us to rethink about pruning in case of reorgs <3
- We need to agree on an interface that should be called from `Ledger` when doing a reorg (something _like_ `undoBlock(block)` )